File: backend/app/state_machine.py
"""Session state machine (Architecture Document §4).

The orchestrator is the only party allowed to move the status, and every
transition is triggered by a clear event.

    SETUP --(start)--> TEACHING --(end)--> ENDED --(evaluate)--> EVALUATED

Rules (§4.2):
- Transitions are forward-only; there is no path back.
- teaching_input is only valid in the TEACHING state.
- Triggering evaluation is only valid in the ENDED state, and is idempotent.
- EVALUATED is terminal.
"""

from __future__ import annotations

from .contracts.enums import SessionStatus

# Events that trigger transitions
START = "start"
END = "end"
EVALUATE = "evaluate"

# Allowed transitions: (current_status, event) -> next_status
_TRANSITIONS: dict[tuple[SessionStatus, str], SessionStatus] = {
    (SessionStatus.SETUP, START): SessionStatus.TEACHING,
    (SessionStatus.TEACHING, END): SessionStatus.ENDED,
    (SessionStatus.ENDED, EVALUATE): SessionStatus.EVALUATED,
}


class InvalidTransition(Exception):
    """A status transition that is not allowed by the state machine."""

    def __init__(self, current: SessionStatus, event: str) -> None:
        self.current = current
        self.event = event
        super().__init__(
            f"Invalid transition: event '{event}' in status '{current}'"
        )


def next_status(current: SessionStatus, event: str) -> SessionStatus:
    """Return the next status for (status, event), or raise InvalidTransition."""
    try:
        return _TRANSITIONS[(SessionStatus(current), event)]
    except KeyError as exc:
        raise InvalidTransition(SessionStatus(current), event) from exc


def can_transition(current: SessionStatus, event: str) -> bool:
    """True if (status, event) is an allowed transition."""
    return (SessionStatus(current), event) in _TRANSITIONS


def accepts_teaching_input(current: SessionStatus) -> bool:
    """teaching_input is only valid in the TEACHING state (§4.2)."""
    return SessionStatus(current) == SessionStatus.TEACHING

File: apps/backend/src/modules/session/stateMachine.ts
/**
 * Session state machine (Architecture Document §4, with an agreed extension).
 *
 * The orchestrator is the only party allowed to move the status, and every
 * transition is triggered by a clear event.
 *
 *   SETUP --(start)--> TEACHING --(end)--> ENDED --(evaluate)--> EVALUATED
 *                          ^                                          |
 *                          +------------------(resume)---------------+
 *
 * Rules:
 * - teaching_input is only valid in the TEACHING state (§4.2).
 * - Triggering evaluation is only valid in the ENDED state, and is idempotent
 *   for the current round (§4.2).
 * - EXTENSION (supersedes the forward-only rule in §4.2, agreed change): a
 *   finished session is resumable — `resume` moves ENDED or EVALUATED back to
 *   TEACHING so the user can continue the conversation. Each ended round still
 *   produces its own EvaluationResult, and the prior results are kept as
 *   history rather than overwritten. EVALUATED is therefore no longer terminal.
 */

import type { SessionStatus } from "../../contracts/enums.js";

// Events that trigger transitions
export const START = "start";
export const END = "end";
export const EVALUATE = "evaluate";
export const RESUME = "resume";

export type SessionEvent =
  | typeof START
  | typeof END
  | typeof EVALUATE
  | typeof RESUME;

// Allowed transitions: `${status}:${event}` -> next status
const TRANSITIONS: Record<string, SessionStatus> = {
  [`SETUP:${START}`]: "TEACHING",
  [`TEACHING:${END}`]: "ENDED",
  [`ENDED:${EVALUATE}`]: "EVALUATED",
  // Resume a finished session to keep teaching (agreed extension).
  [`EVALUATED:${RESUME}`]: "TEACHING",
  [`ENDED:${RESUME}`]: "TEACHING",
};

/** A status transition that is not allowed by the state machine. */
export class InvalidTransition extends Error {
  constructor(
    readonly current: SessionStatus,
    readonly event: string,
  ) {
    super(`Invalid transition: event '${event}' in status '${current}'`);
    this.name = "InvalidTransition";
  }
}

/** Return the next status for (status, event), or throw InvalidTransition. */
export function nextStatus(current: SessionStatus, event: string): SessionStatus {
  const next = TRANSITIONS[`${current}:${event}`];
  if (next === undefined) {
    throw new InvalidTransition(current, event);
  }
  return next;
}

/** True if (status, event) is an allowed transition. */
export function canTransition(current: SessionStatus, event: string): boolean {
  return `${current}:${event}` in TRANSITIONS;
}

/** teaching_input is only valid in the TEACHING state (§4.2). */
export function acceptsTeachingInput(current: SessionStatus): boolean {
  return current === "TEACHING";
}
